app/services/PsrSubmissionService.scala

```diff
@@ -97,29 +100,36 @@ class PsrSubmissionService @Inject()(
       schemeDateService.schemeDate(srn)
     ).mapN { (minimalRequiredSubmission, checkReturnDates, taxYear) =>
       {
-        // TODO fire audit event only if submission is successful
-        val auditEvent = buildAuditEvent(taxYear, loggedInUserNameOrBlank(request))
-        auditService.sendEvent(auditEvent)
-        psrConnector.submitPsrDetails(
-          PsrSubmission(
-            minimalRequiredSubmission = minimalRequiredSubmission,
-            checkReturnDates = checkReturnDates,
-            loans = buildLoans(srn)(optSchemeHadLoans),
-            assets = buildAssets(srn)(
-              optLandOrPropertyHeld,
-              optMoneyWasBorrowed,
-              optDisposeAnyLandOrProperty,
-              optUnregulatedOrConnectedBondsHeld,
-              optBondsDisposal,
-              optOtherAssetsHeld,
-              optOtherAssetsDisposal
-            ),
-            membersPayments = memberPaymentsTransformer.transformToEtmp(srn, request.userAnswers),
-            shares = buildShares(srn)(optDidSchemeHoldAnyShares, optSharesDisposal),
-            psrDeclaration = Option.when(isSubmitted)(declarationTransformer.transformToEtmp)
-          ),
-          optFallbackCall
-        )
+        psrConnector
+          .submitPsrDetails(
+            PsrSubmission(
+              minimalRequiredSubmission = minimalRequiredSubmission,
+              checkReturnDates = checkReturnDates,
+              loans = buildLoans(srn)(optSchemeHadLoans),
+              assets = buildAssets(srn)(
+                optLandOrPropertyHeld,
+                optMoneyWasBorrowed,
+                optDisposeAnyLandOrProperty,
+                optUnregulatedOrConnectedBondsHeld,
+                optBondsDisposal,
+                optOtherAssetsHeld,
+                optOtherAssetsDisposal
+              ),
+              membersPayments = memberPaymentsTransformer.transformToEtmp(srn, request.userAnswers),
+              shares = buildShares(srn)(optDidSchemeHoldAnyShares, optSharesDisposal),
+              psrDeclaration = Option.when(isSubmitted)(declarationTransformer.transformToEtmp)
+            )
+          )
+          .flatMap {
+            case Left(message: String) =>
+              throw PostPsrException(message, optFallbackCall.fold(appConfig.urls.managePensionsSchemes.dashboard) {
```

Comment:
I think optFallbackCall shouldn't be optional. All the controller where used sends fallback url inside Some, We can send it directly, and don't need to fold it later here

Reply:
I'll check how big refactoring it would be.

Reply:
It's a big refactoring, but done it anyway, as we are at the beginning of the sprint.